### Посмотрю на набор данных, проанализирую данные.

In [32]:
import pandas as pd

In [85]:
files_in = pd.read_csv('Input_DIPL/Bookings.csv', sep=";")
# files_in.head(10)
files_in

BookingNumber      Room  TypeRoom Sex  NumberVisits  N_Arrival  \
0              2186       NaN      СЛУЖ   M             2          1   
1              3875      1104      СТД2   M             2          2   
2              2440       604     АПTWN   M             1          1   
3              3438      1014     АПDBL   M             1          1   
4              3805       652     АПDBL   M             1          1   
...             ...       ...       ...  ..           ...        ...   
2526           3930  ВОССТПЛ2  ВОССТПАЛ   F             1          1   
2527           3876      1104      СТД2   M             1          1   
2528           3418      1101      СТД2   F             2          1   
2529           3781      1101      СТД2   F             2          2   
2530           3501       926     АПDBL   F             1          1   

     DateArrival DateDeparture ReservationStatus  DaysInClinic Rate_Passed  \
0     10.01.2019    10.01.2019               OUT             1        ДТ7Б   
1     20.05.2019    21.05.2019               OUT             2      КОРП11   
2     17.03.2019    22.03.2019               OUT             6        ДТ7Б   
3     07.04.2019    13.04.2019               OUT             7        ДТ7Б   
4     01.06.2019    08.06.2019               OUT             8        ДТ7Б   
...          ...           ...               ...           ...         ...   
2526  30.05.2019    31.05.2019               OUT             2    ВОССТПАЛ   
2527  20.05.2019    21.05.2019               OUT             2      КОРП11   
2528  01.04.2019    08.04.2019               OUT             8      КОРП11   
2529  22.05.2019    29.05.2019               OUT             8      КОРП11   
2530  16.06.2019    22.06.2019               OUT             7        ДТ7Б   

        BirthDay  AgeOnArrival                          City  \
0            NaN           200                           NaN   
1            NaN           200                           NaN   
2     10.06.1973            45                        Москва   
3     24.04.1969            49                        Москва   
4     10.04.1965            54                      Прага 10   
...          ...           ...                           ...   
2526  09.07.1986            32                  Петрозаводск   
2527  17.05.1990            29                        Москва   
2528  05.08.1980            38  городское поселение Одинцово   
2529  05.08.1980            38  городское поселение Одинцово   
2530  03.06.1971            48                       Иркутск   

                      Geo  Amount_Bookings  Amount_Additionally  
0     Уточнить при заезде             1960                  971  
1         САНКТ-ПЕТЕРБУРГ             1299                  597  
2                  МОСКВА             9817                  482  
3                  МОСКВА             5067                  244  
4                   Чехия              344                  428  
...                   ...              ...                  ...  
2526  Уточнить при заезде             6408                  832  
2527      САНКТ-ПЕТЕРБУРГ             7382                  347  
2528               МОСКВА              160                  380  
2529               МОСКВА             7449                   80  
2530       ИРКУТСКАЯ ОБЛ.             8357                  262  

[2531 rows x 17 columns]

### Описания полей:

1. BookingNumber - номер брони
2. Room - номер комнаты
3. TypeRoom - тип комнаты
4. Sex - Пол
5. NumberVisits - кол-во визитов 
6. N_Arrival - номер заезда
7. DateArrival - дата заезда
8. DateDeparture - дата выезда
9. ReservationStatus - статус брони
10. DaysInClinic - дней в клинике
11. Rate_Passed - программа
12. BirthDay - день рождения гостя
13. AgeOnArrival - возраст на момент заезда
14. City - город гостя
15. Geo - гео локация гостя
16. Amount_Bookings - сумма брони
17. Amount_Additionally - сумма доп. услуг

In [86]:
files_in.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2531 entries, 0 to 2530
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   BookingNumber        2531 non-null   int64 
 1   Room                 2526 non-null   object
 2   TypeRoom             2531 non-null   object
 3   Sex                  2531 non-null   object
 4   NumberVisits         2531 non-null   int64 
 5   N_Arrival            2531 non-null   int64 
 6   DateArrival          2531 non-null   object
 7   DateDeparture        2531 non-null   object
 8   ReservationStatus    2531 non-null   object
 9   DaysInClinic         2531 non-null   int64 
 10  Rate_Passed          2531 non-null   object
 11  BirthDay             2527 non-null   object
 12  AgeOnArrival         2531 non-null   int64 
 13  City                 2406 non-null   object
 14  Geo                  2531 non-null   object
 15  Amount_Bookings      2531 non-null   int64 
 16  Amount

### Всего в наборе 2531 запись и 17 полей.
### Целочисленных: 6 полей, полей объектов: 10

####  В наборе имеются пропуски (Null) по полям: Room (5 записей), BirthDay (4 записи), City (125 записей)

### Посмотрю на данные подробнее:

In [54]:
print(f'Уникальные значения: {files_in["BookingNumber"].unique()}')
print(f'Количество уникальных значений: {len(files_in["BookingNumber"].unique())}')

Уникальные значения: [2186 3875 2440 ... 3418 3781 3501]
Количество уникальных значений: 2531


#### то есть все значения в поле BookingNumber - уникальны.

Посмотрю на остальные поля:

In [55]:
print(f'Уникальные значения: {files_in["Room"].unique()}')
print(f'Количество уникальных значений: {len(files_in["Room"].unique())}')

Уникальные значения: [nan '1104' '604' '1014' '652' '257' '258' '207' '664' '255' '612' '314'
 '864' '1012' '253' '753' '211' '1062' '215' '353' '355' '811' '662' '203'
 '301' '303' '210' '1056' '614' '350' '926' '307' '213' '953' '305' '602'
 '1003' '917' '752' '611' '656' '912' '911' '660' '814' '859' '1006'
 '1064' '1063' '261' '1051' '309' '610' '920' '703' '918' '667' '351'
 '919' '654' '1055' '1057' '914' '1053' '952' '259' '1054' '951' '254'
 '927' '701' '704' '1010' '251' '1001' '1058' '862' '751' '606' '812'
 '1102' '702' '661' '212' '1015' '1016' '357' '930' '1008' '668' '1101'
 '616' '1061' '209' '617' '1059' '1103' '1106' '754' '1108' 'ВОССТПЛ2'
 '1009' 'ПЕРЕСЕЛ' '1007' 'ВОССТПАЛ']
Количество уникальных значений: 105


In [56]:
print(f'Уникальные значения: {files_in["TypeRoom"].unique()}')
print(f'Количество уникальных значений: {len(files_in["TypeRoom"].unique())}')

Уникальные значения: ['СЛУЖ' 'СТД2' 'АПTWN' 'АПDBL' 'АПSGL' 'СЮИТ' 'АПTWNT' 'СЮИТФ' 'АПDBLCLS'
 'АПФ' 'СТУД' 'СЮИТПР' 'СЮИТDLX' 'СЮИТТРС' 'СТД1' 'ВОССТПАЛ']
Количество уникальных значений: 16


In [57]:
print(f'Уникальные значения: {files_in["Sex"].unique()}')
print(f'Количество уникальных значений: {len(files_in["Sex"].unique())}')

Уникальные значения: ['M' 'F']
Количество уникальных значений: 2


In [58]:
print(f'Уникальные значения: {files_in["NumberVisits"].unique()}')
print(f'Количество уникальных значений: {len(files_in["NumberVisits"].unique())}')

Уникальные значения: [2 1 3 6]
Количество уникальных значений: 4


In [59]:
print(f'Уникальные значения: {files_in["N_Arrival"].unique()}')
print(f'Количество уникальных значений: {len(files_in["N_Arrival"].unique())}')

Уникальные значения: [1 2 3 4 5 6]
Количество уникальных значений: 6


In [60]:
print(f'Уникальные значения: {files_in["DateArrival"].unique()}')
print(f'Количество уникальных значений: {len(files_in["DateArrival"].unique())}')

Уникальные значения: ['10.01.2019' '20.05.2019' '17.03.2019' '07.04.2019' '01.06.2019'
 '10.02.2019' '24.02.2019' '24.03.2019' '13.01.2019' '12.05.2019'
 '29.05.2019' '28.04.2019' '03.03.2019' '19.05.2019' '03.01.2019'
 '09.06.2019' '09.02.2019' '27.01.2019' '06.04.2019' '05.05.2019'
 '10.03.2019' '14.04.2019' '23.06.2019' '17.04.2019' '31.03.2019'
 '16.06.2019' '21.04.2019' '03.02.2019' '17.02.2019' '13.02.2019'
 '20.04.2019' '19.01.2019' '20.01.2019' '09.05.2019' '02.06.2019'
 '26.05.2019' '07.02.2019' '24.04.2019' '06.03.2019' '19.06.2019'
 '08.06.2019' '01.05.2019' '08.05.2019' '08.02.2019' '16.02.2019'
 '27.03.2019' '09.03.2019' '20.03.2019' '23.02.2019' '29.01.2019'
 '15.01.2019' '10.04.2019' '11.03.2019' '18.05.2019' '23.01.2019'
 '24.01.2019' '10.05.2019' '16.03.2019' '17.01.2019' '11.05.2019'
 '04.01.2019' '06.01.2019' '01.04.2019' '22.05.2019' '21.01.2019'
 '04.02.2019' '14.02.2019' '28.02.2019' '24.06.2019' '20.06.2019'
 '30.03.2019' '03.04.2019' '20.02.2019' '23.04.2019' '2

In [61]:
print(f'Уникальные значения: {files_in["DateDeparture"].unique()}')
print(f'Количество уникальных значений: {len(files_in["DateDeparture"].unique())}')

Уникальные значения: ['10.01.2019' '21.05.2019' '22.03.2019' '13.04.2019' '08.06.2019'
 '16.02.2019' '02.03.2019' '30.03.2019' '19.01.2019' '18.05.2019'
 '07.06.2019' '07.05.2019' '09.03.2019' '28.05.2019' '12.01.2019'
 '15.06.2019' '23.02.2019' '23.03.2019' '09.02.2019' '05.03.2019'
 '16.03.2019' '27.04.2019' '06.07.2019' '24.04.2019' '05.05.2019'
 '06.04.2019' '22.06.2019' '11.05.2019' '20.04.2019' '12.02.2019'
 '22.02.2019' '16.01.2019' '26.01.2019' '25.01.2019' '17.05.2019'
 '12.03.2019' '13.07.2019' '10.05.2019' '29.06.2019' '16.06.2019'
 '18.01.2019' '21.03.2019' '10.03.2019' '15.02.2019' '12.04.2019'
 '03.05.2019' '04.05.2019' '02.02.2019' '22.01.2019' '18.06.2019'
 '15.03.2019' '28.06.2019' '26.02.2019' '20.02.2019' '01.06.2019'
 '18.03.2019' '06.02.2019' '08.04.2019' '07.03.2019' '29.03.2019'
 '17.02.2019' '27.03.2019' '16.04.2019' '27.06.2019' '08.03.2019'
 '26.04.2019' '09.06.2019' '26.05.2019' '30.04.2019' '05.04.2019'
 '25.05.2019' '19.06.2019' '11.06.2019' '23.01.2019' '1

In [62]:
print(f'Уникальные значения: {files_in["ReservationStatus"].unique()}')
print(f'Количество уникальных значений: {len(files_in["ReservationStatus"].unique())}')

Уникальные значения: ['OUT']
Количество уникальных значений: 1


In [63]:
print(f'Уникальные значения: {files_in["DaysInClinic"].unique()}')
print(f'Количество уникальных значений: {len(files_in["DaysInClinic"].unique())}')

Уникальные значения: [ 1  2  6  7  8 10 15 14 52 35 21 17 19 20 11  9 12 13  5  4  3 18 16 22
 28]
Количество уникальных значений: 25


In [64]:
print(f'Уникальные значения: {files_in["Rate_Passed"].unique()}')
print(f'Количество уникальных значений: {len(files_in["Rate_Passed"].unique())}')

Уникальные значения: ['ДТ7Б' 'КОРП11' 'ДТ7Э' 'ДТ10Б' 'ООР7П' 'ДТ10П' 'ООР10П' 'АНСТР10'
 'ДБ14АВИА' 'ООР14Б' 'ООР7Б' 'ДТ14П' 'ДЕТЗДР14' 'ДТ14Б' 'ДТ14Э' 'ЗДОР60+'
 'ООР10Б' 'ППЖ14' 'ДБ7АВИА' 'ДТ7П' 'ПП10' 'ПП14' 'ПП21' 'ПП' 'ППМ10'
 'ДЕТЗДР7' 'ЛГ21' 'АНСТР14' 'АНСТР7' 'ДЕТЗДР10' 'ЛГ14' 'ДТ21Б' 'ДТ21П'
 'ООР21Б' 'ООР14П' 'ЗД60+10Д' 'ВОССТПАЛ' 'ЛЕЧБЕСПМ' 'ДТ10Э' 'ООР21П'
 'ДП14АВИА' 'КОРП11ДН' 'ОБ10АВИА' 'ЛЕЧБЕСПЖ' 'СЛУЖ']
Количество уникальных значений: 45


In [65]:
print(f'Уникальные значения: {files_in["BirthDay"].unique()}')
print(f'Количество уникальных значений: {len(files_in["BirthDay"].unique())}')

Уникальные значения: [nan '10.06.1973' '24.04.1969' ... '17.05.1990' '05.08.1980' '03.06.1971']
Количество уникальных значений: 2276


In [66]:
print(f'Уникальные значения: {files_in["AgeOnArrival"].unique()}')
print(f'Количество уникальных значений: {len(files_in["AgeOnArrival"].unique())}')

Уникальные значения: [200  45  49  54  52  62  40  58  65  50  60  48  47  41  66  64  57  38
  21  51  12  61  63  44  56  32  46  59  17  42  31  37  55  53  39  10
  43  19  36  69  34  22   5   8  28  76  33  67  35   9  68  18  29  70
  27  30  14  13  20  26  15  16   3   7  11  75  80  72   2   6  25  23
  24  82   1   0   4  71  78  73  84  74  77]
Количество уникальных значений: 83


In [81]:
print(f'Количество записей с AgeOnArrival=200: {len([item for item in files_in["AgeOnArrival"] if item == 200])}')

Количество записей с AgeOnArrival=200: 4


### В данном случае возраст: 200 не может быть.
## Буду строки с этим возрастом выкидывать.
#### Возраст 200 был установлен мной при выгрузке из БД для тех гостей по которым не указана дата рождения.

In [67]:
print(f'Уникальные значения: {files_in["City"].unique()}')
print(f'Количество уникальных значений: {len(files_in["City"].unique())}')

Уникальные значения: [nan 'Москва' 'Прага 10' 'Калуга' 'Дружба' 'Орехово-Зуево'
 'Санкт-Петербург' 'Сургут' 'Кромино' 'г. Москва' 'Москва ' 'Екатеринбург'
 'Смоленск' 'Сертолово' 'Мурманск' 'МОСКВА' 'Сычевичи' 'Нижний Новгород'
 'Химки, квартал Клязьма ' 'Тамбов' 'Уфа' 'Ямное' 'г. Санкт-Петербург'
 'Атяшевский р-н' ' Москва' 'Петрозаводск' 'Пенза' 'Воронеж' 'Астана'
 'Токсово' 'Ставрополь' 'город Москва' 'город Москва ' 'Краснодар'
 'Краснотурьинск г' 'Успенское' 'Ташкент' 'Великие Луки' 'Минск' 'Аксай'
 'Калининград' 'Калиниград' 'д. Кузьминка' 'г. Мурманск' 'Волгоград'
 'Новый Уренгой' 'г.Химки' 'Нарьян-Мар' 'г. Сосновый Бор'
 'город Нижний Новгород' 'Геленджик' 'Дзержинский' 'Свердловск'
 'гор. Москва' 'Новороссийск' 'Янино-1 дер.' 'Бугульма'
 'дер.Новоглаголево' 'Томск' 'Подольск' 'Марфино' 'Подольск ' 'Саловка'
 'Буденновск' 'Курск' 'Запорожье' 'Дроздово' 'Апрелевка' 'Вольск'
 ' город Москва' 'Вильнюс' 'Химки' 'Москва г' 'Бугры' 'Казань'
 'г. Москва ' 'Чайковский' 'Ростов-на-Дону'

In [68]:
print(f'Уникальные значения: {files_in["Geo"].unique()}')
print(f'Количество уникальных значений: {len(files_in["Geo"].unique())}')

Уникальные значения: ['Уточнить при заезде' 'САНКТ-ПЕТЕРБУРГ' 'МОСКВА' 'Чехия' 'КАЛУЖСКАЯ ОБЛ.'
 'АЛТАЙСКИЙ КРАЙ' 'МОСКОВСКАЯ ОБЛ.' 'ХАНТЫ-МАНСИЙСКИЙ АО' 'Армения'
 'Иностранцы' 'СМОЛЕНСКАЯ ОБЛ.' 'МУРМАНСКАЯ ОБЛ.' 'КАЗАХСТАН' 'Белоруссия'
 'НИЖЕГОРОДСКАЯ ОБЛ.' 'БАШКОРТОСТАН' 'ВОРОНЕЖСКАЯ ОБЛ.' 'МОРДОВИЯ'
 'КАРЕЛИЯ' 'ЛЕНИНГРАДСКАЯ ОБЛ.' 'ПЕНЗЕНСКАЯ ОБЛ.' 'СТАВРОПОЛЬСКИЙ КРАЙ'
 'ТАМБОВСКАЯ ОБЛ.' 'КРАСНОДАРСКИЙ КРАЙ' 'Латвия' 'Узбекистан' 'США'
 'НОВГОРОДСКАЯ ОБЛ.' 'МАРИЙ ЭЛ' 'КАЛИНИНГРАДСКАЯ ОБЛ.'
 'ВОЛГОГРАДСКАЯ ОБЛ.' 'ЯМАЛО-НЕНЕЦКИЙ АО' 'Испания' 'СВЕРДЛОВСКАЯ ОБЛ.'
 'ТАТАРСТАН' 'САРАТОВСКАЯ ОБЛ.' 'КУРСКАЯ ОБЛ.' 'Украина' 'РОСТОВСКАЯ ОБЛ.'
 'УДМУРТИЯ' 'БЕЛГОРОДСКАЯ ОБЛ.' 'АРХАНГЕЛЬСКАЯ ОБЛ.' 'ВОЛОГОДСКАЯ ОБЛ.'
 'ЧУКОТСКИЙ АО' 'ПЕРМСКАЯ ОБЛ.' 'ОМСКАЯ ОБЛ.' 'ЛИПЕЦКАЯ ОБЛ.'
 'ХАБАРОВСКИЙ КРАЙ' 'КОМИ' 'ЧЕЛЯБИНСКАЯ ОБЛ.' 'ПРИМОРСКИЙ КРАЙ'
 'ИРКУТСКАЯ ОБЛ.' 'РЯЗАНСКАЯ ОБЛ.' 'ТВЕРСКАЯ ОБЛ.' 'САХАЛИНСКАЯ ОБЛ.'
 'ОРЕНБУРГСКАЯ ОБЛ.' 'ТУЛЬСКАЯ ОБЛ.' 'Швеция' 'САМАРСКАЯ ОБЛ.'
 'ЯРОСЛАВСКАЯ ОБЛ.' '

In [89]:
print(f'Уникальные значения: {files_in["Amount_Bookings"].unique()}')
print(f'Количество уникальных значений: {len(files_in["Amount_Bookings"].unique())}')

Уникальные значения: [1960 1299 9817 ... 7382  160 7449]
Количество уникальных значений: 2219


In [87]:
print(f'Уникальные значения: {files_in["Amount_Additionally"].unique()}')
print(f'Количество уникальных значений: {len(files_in["Amount_Additionally"].unique())}')

Уникальные значения: [ 971  597  482  244  428  854  883  102  301  882  260   16  449  156
  229   48  587  644  172  696  388  843  604  708  154  799   18  469
  440  485   38  926   21  759  236  245  584    3  124  435  778  533
  608  458  621  178  324  396  782  280  806   34  719  715  508  956
  598  275  493  288  775  402  974   82  527  668  707  947   81  546
   47  702  817  418  295  581  949  521  638   14  787  611  935  880
  781  783  555  421  480  315  393  392  296  498  870  825  695  222
  972  661  262  845  740  742  809   74  307  762  169  526  942  196
   54  620  769  826  115  833  121  968  713  225  865  108   51  474
  884  776  800  941   91  487  623  497   60  382  309  276  313  605
  540  635  334  914  803  774  386  590  550  744  679  265   88   15
  544  234  534  358  495  372  133  343  233  700  984  626   50  319
  212  950  763  104  164   33  442  636  602   39  368  886  249  345
  828  693  728  451  677  328   89  551  144  130    9 

### В целом с набором данных не так плохо. Надо будет выкидывать строки с Null значениями и возраст = 200